### ISSUE TITLE
Timeout too short for (some?) DHT11 sensors

### ISSUE THREAD

**Issue by jamboni** [RobTillaart/Arduino#61]
TL;DR
I couldn't get my DHT11 to work without increasing the DHTLIB_TIMEOUT by a factor of 100

-------
The SODIAL 3 pin DHT11's I purchased from Amazon wouldn't work with the library as is. I received "connect error"s when running the DHT11 test program.

Specific sensor: https://www.amazon.ca/gp/product/B00K67YK6M/ref=oh_aui_search_detailpage?ie=UTF8&psc=1

After watching a working DHT22 alongside the DHT11 with my oscilloscope I noticed the response time for the DHT22 was much faster than the DHT11.

I opened `dht.h` and increased the delay by a factor of 100. (10 wasn't enough). `DHTLIB_TIMEOUT` ended up being set at 40000.

    #ifndef F_CPU
    //#define DHTLIB_TIMEOUT 1000  // ahould be approx. clock/40000
    #define DHTLIB_TIMEOUT 10000  // ahould be approx. clock/40000
    #else
    //#define DHTLIB_TIMEOUT (F_CPU/40000)
    #define DHTLIB_TIMEOUT (F_CPU/400)
    #endif

After that I was able to get readings.

**Comment by jigten** [comment]
I was also stuck with this problem for over 3 weeks. After changing the value to 400 I was able to finally get the readings. Thank you for posting this here. May I ask what board you are using? I was using the Arduino Pro Mini and maybe this problem is specific to this board?

**Comment by jamboni** [comment]
I'm using an UNO Rev3 (specifically [https://www.amazon.ca/gp/product/B00KJGLTCI/](https://www.amazon.ca/gp/product/B00KJGLTCI/))

I assume it's more likely an issue with the DHT library, or the DHT11 sensors than an issue with the board. The DHT11's simply won't reply fast enough for the code.

According to the [datasheet](http://www.micropik.com/PDF/dht11.pdf) for the DHT11 (see section 5.2), it should only take 20-40us to get a response.

Spec's listed on this page [arduino-info.wikispaces.com/DHT11-Humidity-TempSensor](https://arduino-info.wikispaces.com/DHT11-Humidity-TempSensor). Which appear to have been amended by Rob Tillaart (the developer of this lib) suggest that the response time could be up to 50ms.

I should have noted the actual response with my oscilloscope while I had everything set up, but it's all disassembled now. I wouldn't be surprised if it was as much as 300-400ms.

**Comment by jigten** [comment]
I see, so from what I understand the code did not allow the micro-controller to wait long enough for a response to be actually received from the Arduino? I'm very new to Arduino at the moment so I am just trying out some beginner projects and didn't expect to get this much trouble from what I believed would be a simple and easy to implement temperature sensor.  So by changing the value to 400 we effectively raised the time out period from 100us to 10 ms? So is the datasheet is wrong when it reports that the response will take 20-40us? Im sorry if these are trivial questions but I'm just trying to better understand why this problem occured and why this change in the code fixed it. Thank you for your kind help and time. Any information you can provide will be greatly appreciated.

**Comment by jamboni** [comment]
Exactly.

In dht.cpp, around line 130, in the readSensor() function there is a while loop that runs for at most DHTLIB_TIMEOUT*2 iterations. DHTLIB_ERROR_CONNECT is thrown if no response is received by the time the loop ends. The loop is intended to run for no more than 200us, what we did was bump that up closer to 20ms.

I hooked everything up again, my oscilloscope shows that between the Arduino sending the read request and receiving a response, it is taking between 6 to 8ms.

I'm not sure if it's an issue with the datasheet or perhaps we bought DHT11's that don't work the way they are supposed to. This is how all of the DHT11's I have are behaving, maybe someone else can let us know if they have a DHT11 that works without modifying the timeout. The DHT22 I have works with the 200us timeout.

**Comment by RobTillaart** [comment]
Sorry for the late response,
I wrote the lib with using a DHT11 and DHT22 bought from a local supplier in the Netherlands. Both behaved according to specification.

**Comment by jigten** [comment]
Thank you for explaining. I have gotten a much more clear understanding of this problem.

**Comment by ogiewon** [comment]
I have used both DHT-11's and DHT-22's with Arduino UNO R3's a MEGA 2560's without modifying Rob's library in any way.  My DHT-11's were bought years ago though.  The DHT-22's are all I use these days and recently purchased ones from eBay/China work fine.  I did have to go back to the "Stable" release for them to work with the ESP8266 microcontrollers.
